cli/exchange/node.go

```diff
@@ -465,3 +500,292 @@ func NodeListErrors(org string, credToUse string, node string, long bool) {
+	// if node has a pattern, check the service is defined in top level services
+	nodes := exNode.Nodes
+	if len(nodes) != 1 {
+		errorString = msgPrinter.Sprintf("Expecting 1 exchange node, but get %d nodes", len(nodes))
+		return false, warningMessage, errors.New(errorString)
+	}
+
+	nodePatternString := ""
+	for _, node := range nodes {
+		nodePatternString = node.Pattern
+	}
+
+	// Need to check:
+	// 1) if service is top level service
+	// 2) or recursively check if service defined in dependent service
+	// 3) if service range of the dependent service has intersection with versionRange of service from userInput
+	if nodePatternString != "" {
+		patOrg, patternName := cliutils.TrimOrg(org, nodePatternString)
+		var patternsResp ExchangePatterns
+		// Get exchange pattern
+		httpCode := cliutils.ExchangeGet("Exchange", cliutils.GetExchangeUrl(), "orgs/"+patOrg+"/patterns"+cliutils.AddSlash(patternName), cliutils.OrgAndCreds(org, credToUse), []int{200, 404}, &patternsResp)
+		if httpCode == 404 {
+			errorString = msgPrinter.Sprintf("Pattern %v of the node does not exist for org: %v \n", nodePatternString, patOrg)
+			return false, warningMessage, errors.New(errorString)
+		}
+
+		patterns := patternsResp.Patterns
+		if len(patterns) != 1 {
+			errorString = msgPrinter.Sprintf("Expecting 1 pattern with name %v, but get %d patterns", nodePatternString, len(nodes))
+			return false, warningMessage, errors.New(errorString)
+		}
+
+		pattern, _ := patterns[nodePatternString]
+		topLevelServices := pattern.Services
+		for _, topLevelService := range topLevelServices {
+			var topLevelServiceVersion ServiceChoice
+			if serviceOrg == topLevelService.ServiceOrg && serviceArch == topLevelService.ServiceArch && serviceUrl == topLevelService.ServiceURL {
+				for _, topLevelServiceVersion = range topLevelService.ServiceVersions {
+					if withinRange, _ := vExp.Is_within_range(topLevelServiceVersion.Version); withinRange == true {
+						return true, warningMessage, nil
+					}
+				}
+			}
+
+			// recursively check required services/dependent services
+			servToCheckDependentService := topLevelService
+			version := topLevelServiceVersion.Version
+			versionRange := ""
+			findMatchInDepService := checkDependentServices(cliutils.OrgAndCreds(org, credToUse), serviceOrg, serviceArch, serviceUrl, vExp, servToCheckDependentService, version, versionRange)
+			if findMatchInDepService {
+				return true, warningMessage, nil
+			}
+
+		}
+
+		errorString = msgPrinter.Sprintf("Service %v %v %v %v in userInput is not defined in pattern or dependent services", serviceOrg, serviceUrl, serviceArch, vExp.Get_expression())
+		return false, warningMessage, errors.New(errorString)
+
+	}
+	return true, warningMessage, nil
+
+}
+
+func checkDependentServices(creds string, serviceOrg string, serviceArch string, serviceUrl string, serviceVersionExp *semanticversion.Version_Expression, servToCheckDependentService ServiceReference, version string, versionRange string) bool {
```

Comment:
Not sure where the problem is. If a node is using a pattern, and the top level service requires another service, and the required service has userinputs without default values, then we need to catch that case and issue an error.

There is also a related problem that is even more simple to create. If the top level service has a userinput with no default value AND that userinput is not set in the input body, that should also be flagged as an error.

Reply:
The second problem seems to be fixed, but the first problem is not. A node with a pattern, where the pattern's top level service requires another service that has userinputs but no default values, is not flagged as in error.